libknet/crypto_openssl.c

```diff
@@ -395,12 +395,96 @@ static int opensslcrypto_authenticate_and_decrypt (
 	return 0;
 }
 
+#ifdef BUILDCRYPTOOPENSSL10
+static pthread_mutex_t *openssl_internal_lock;
+static long *openssl_internal_lock_count;
+
+static void openssl_internal_locking_callback(int mode, int type, char *file, int line)
+{
+	if (mode & CRYPTO_LOCK) {
+		pthread_mutex_lock(&(openssl_internal_lock[type]));
+		openssl_internal_lock_count[type]++;
+	} else {
+		pthread_mutex_unlock(&(openssl_internal_lock[type]));
+	}
+}
+
+static unsigned long openssl_internal_thread_id(void)
+{
+	return (unsigned long)pthread_self();
+}
+
+static void openssl_internal_lock_cleanup(void)
+{
+	int i;
+
+	CRYPTO_set_locking_callback(NULL);
+	CRYPTO_set_id_callback(NULL);
+
+	for (i = 0; i < CRYPTO_num_locks(); i++) {
+		pthread_mutex_destroy(&(openssl_internal_lock[i]));
+	}
+
+	if (openssl_internal_lock_count) {
+		free(openssl_internal_lock_count);
+	}
+
+	if (openssl_internal_lock) {
+		free(openssl_internal_lock);
+	}
+
+	return;
+}
+
+static int openssl_internal_lock_setup(void)
+{
+	int savederrno = 0, err = 0;
+	int i;
+
+	openssl_internal_lock = malloc(CRYPTO_num_locks() * sizeof(pthread_mutex_t));
+	if (!openssl_internal_lock) {
+		savederrno = errno;
+		err = -1;
+		goto out;
+	}
+
+	openssl_internal_lock_count = malloc(CRYPTO_num_locks() * sizeof(long));
+	if (!openssl_internal_lock_count) {
+		savederrno = errno;
+		err = -1;
+		goto out;
+	}
+
+	for (i = 0; i < CRYPTO_num_locks(); i++) {
+		openssl_internal_lock_count[i] = 0;
+		savederrno = pthread_mutex_init(&(openssl_internal_lock[i]), NULL);
+		if (savederrno) {
+			err = -1;
+			goto out;
+		}
+	}
+
+	CRYPTO_set_id_callback((unsigned long (*)(void))openssl_internal_thread_id);
```

Comment:
Hi Fabio, the OpenSSL [documentation page](https://www.openssl.org/docs/man1.0.2/crypto/threads.html) you referenced calls `CRYPTO_set_id_callback()` deprecated, because it assumes that thread IDs can always be represented by `unsigned long`. Even if it's true for BSDs and Linux on mainstream architectures, are you sure you want to buy into that?
Regards,
Feri.

Reply:
I simply copied the code provided with openssl-1.0.2 in their testsuite and referenced around. It's unclear if the documentation is uptodate based on the blog mentioned in the commit. Even if it's deprecated, it does the job and the whole test suite is passing. Given that openssl < 1.1 is also on the way out (very few distro still ship it), I am happy to leave it as-is.